File: go/solutions/day01.go
package solutions

import (
	"sort"
	"strconv"
	"strings"
)

type Day01 struct{}

func (d *Day01) First(input string) int {
	var left []string
	var right []string

	for _, line := range strings.Split(input, "\n") {
		parts := strings.Fields(line)
		if len(parts) >= 2 {
			left = append(left, strings.TrimSpace(parts[0]))
			right = append(right, strings.TrimSpace(parts[1]))
		}
	}

	sort.Strings(left)
	sort.Strings(right)

	sum := 0
	for i := 0; i < len(left) && i < len(right); i++ {
		leftNum, err1 := strconv.Atoi(left[i])
		rightNum, err2 := strconv.Atoi(right[i])

		if err1 == nil && err2 == nil {
			sum += abs(leftNum - rightNum)
		}
	}

	return sum
}

func (d *Day01) Second(input string) int {
	leftSet := make(map[string]struct{})
	rightCounts := make(map[string]int)

	for _, line := range strings.Split(input, "\n") {
		parts := strings.Fields(line)
		if len(parts) >= 2 {
			left := strings.TrimSpace(parts[0])
			right := strings.TrimSpace(parts[1])

			leftSet[left] = struct{}{}

			rightCounts[right]++
		}
	}

	sum := 0
	for left := range leftSet {
		if count, exists := rightCounts[left]; exists {
			leftValue, err := strconv.Atoi(left)
			if err == nil {
				sum += leftValue * count
			}
		}
	}

	return sum
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: rust/src/solutions/day_01.rs
use std::collections::{HashMap, HashSet};

use crate::solvable::Solvable;

pub struct Day01;

impl Solvable for Day01 {
    fn first(&self, input: &str) -> i32 {
        let input = input
            .lines()
            .filter_map(|line| {
                if let Some((left, right)) = line.split_once(char::is_whitespace) {
                    Some((left.trim().to_string(), right.trim().to_string()))
                } else {
                    None
                }
            })
            .collect::<(Vec<String>, Vec<String>)>();
        let (mut left, mut right) = input;

        left.sort();
        right.sort();

        left.into_iter()
            .zip(right)
            .into_iter()
            .map(|(left, right)| {
                (left.parse::<i32>())
                    .unwrap()
                    .abs_diff(right.parse::<i32>().unwrap()) as i32
            })
            .sum()
    }

    fn second(&self, input: &str) -> i32 {
        let input = input
            .lines()
            .filter_map(|line| {
                if let Some((left, right)) = line.split_once(char::is_whitespace) {
                    Some((left.trim().to_string(), right.trim().to_string()))
                } else {
                    None
                }
            })
            .collect::<(HashSet<_>, Vec<_>)>();
        let (left, right) = input;

        let right = right.into_iter().fold(HashMap::new(), |mut acc, item| {
            *acc.entry(item).or_insert(0) += 1;
            acc
        });

        left.into_iter()
            .map(|value| match right.get(&value) {
                Some(mul) => value.parse::<i32>().unwrap() * *mul,
                None => 0,
            })
            .sum::<i32>()
    }
}
